test/integration/targets/copy/tasks/tests.yml

```diff
@@ -531,6 +531,7 @@
     directory_mode: 0700
     local_follow: False
   register: recursive_copy_result
+  diff: true
```

Comment:
Doesn't this succeed with or without the patch? The test should validate the fix.

Reply:
I'm not sure exactly what assertion I should make to validate this. Is it sufficient to assert that result.diff is a list and that it has a length > 1?

This is the output:
<details>

```console
--- before
+++ after: /work/pi_tbernard_umass_edu/simonleary_umass_edu/ansible/ansible-official-simon/test/results/.tmp/integration/copy-72dyldkz-ÅÑŚÌβŁÈ/test/integration/targets/copy/files/subdir/bar.txt
@@ -0,0 +1 @@
+baz

--- before
+++ after: /work/pi_tbernard_umass_edu/simonleary_umass_edu/ansible/ansible-official-simon/test/results/.tmp/integration/copy-72dyldkz-ÅÑŚÌβŁÈ/test/integration/targets/copy/files/subdir/subdir2/baz.txt
@@ -0,0 +1 @@
+baz

--- before
+++ after: /work/pi_tbernard_umass_edu/simonleary_umass_edu/ansible/ansible-official-simon/test/results/.tmp/integration/copy-72dyldkz-ÅÑŚÌβŁÈ/test/integration/targets/copy/files/subdir/subdir2/subdir3/subdir4/qux.txt
@@ -0,0 +1 @@
+qux
\ No newline at end of file

--- before
+++ after
@@ -1,5 +1,5 @@
 {
-    "mode": "0775",
+    "mode": "0700",
     "path": "/tmp/ansible.pn20_lyf.test/copy/sub/subdir/subdira",
-    "state": "absent"
+    "state": "directory"
 }

--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "/tmp/ansible.pn20_lyf.test/copy/sub/subdir/subdir1/ansible-test-abs-link",
-    "state": "absent"
+    "state": "link"
 }

--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "/tmp/ansible.pn20_lyf.test/copy/sub/subdir/subdir1/bar.txt",
-    "state": "absent"
+    "state": "link"
 }

--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "/tmp/ansible.pn20_lyf.test/copy/sub/subdir/subdir1/invalid",
-    "state": "absent"
+    "state": "link"
 }

--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "/tmp/ansible.pn20_lyf.test/copy/sub/subdir/subdir1/invalid2",
-    "state": "absent"
+    "state": "link"
 }

--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "/tmp/ansible.pn20_lyf.test/copy/sub/subdir/subdir1/out_of_tree_circle",
-    "state": "absent"
+    "state": "link"
 }

--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "/tmp/ansible.pn20_lyf.test/copy/sub/subdir/subdir1/ansible-test-abs-link-dir",
-    "state": "absent"
+    "state": "link"
 }

--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "/tmp/ansible.pn20_lyf.test/copy/sub/subdir/subdir1/circles",
-    "state": "absent"
+    "state": "link"
 }

--- before
+++ after
@@ -1,4 +1,4 @@
 {
     "path": "/tmp/ansible.pn20_lyf.test/copy/sub/subdir/subdir1/subdir3",
-    "state": "absent"
+    "state": "link"
 }
```

</details>

I was going to make an assertion that `diff[0]` is `state=absent -> state=directory`, and all following diff elements are `state=absent -> state=link`, but that's not actually true. Here is the actual result object:

<details>

```json
    "recursive_copy_result": {
        "changed": true,
        "dest": "/tmp/ansible.gdf5vnzz.test/copy/sub/",
        "diff": [
            {
                "after": "baz\n",
                "after_header": "/Users/simon/ansible/test/results/.tmp/integration/copy-14vikdc8-ÅÑŚÌβŁÈ/test/integration/targets/copy/files/subdir/bar.txt",
                "before": ""
            },
            {
                "after": "",
                "after_header": "/Users/simon/ansible/test/results/.tmp/integration/copy-14vikdc8-ÅÑŚÌβŁÈ/test/integration/targets/copy/files/subdir/subdir1/empty.txt",
                "before": ""
            },
            {
                "after": "baz\n",
                "after_header": "/Users/simon/ansible/test/results/.tmp/integration/copy-14vikdc8-ÅÑŚÌβŁÈ/test/integration/targets/copy/files/subdir/subdir2/baz.txt",
                "before": ""
            },
            {
                "after": "qux",
                "after_header": "/Users/simon/ansible/test/results/.tmp/integration/copy-14vikdc8-ÅÑŚÌβŁÈ/test/integration/targets/copy/files/subdir/subdir2/subdir3/subdir4/qux.txt",
                "before": ""
            },
            {
                "after": {
                    "path": "/tmp/ansible.gdf5vnzz.test/copy/sub/subdir/subdir1"
                },
                "before": {
                    "path": "/tmp/ansible.gdf5vnzz.test/copy/sub/subdir/subdir1"
                }
            },
            {
                "after": {
                    "path": "/tmp/ansible.gdf5vnzz.test/copy/sub/subdir/subdir2"
                },
                "before": {
                    "path": "/tmp/ansible.gdf5vnzz.test/copy/sub/subdir/subdir2"
                }
            },
            {
                "after": {
                    "mode": "0700",
                    "path": "/tmp/ansible.gdf5vnzz.test/copy/sub/subdir/subdira",
                    "state": "directory"
                },
                "before": {
                    "mode": "0775",
                    "path": "/tmp/ansible.gdf5vnzz.test/copy/sub/subdir/subdira",
                    "state": "absent"
                }
            },
            {
                "after": {
                    "path": "/tmp/ansible.gdf5vnzz.test/copy/sub/subdir/subdir2/subdir3"
                },
                "before": {
                    "path": "/tmp/ansible.gdf5vnzz.test/copy/sub/subdir/subdir2/subdir3"
                }
            },
            {
                "after": {
                    "path": "/tmp/ansible.gdf5vnzz.test/copy/sub/subdir/subdir2/subdir3/subdir4"
                },
                "before": {
                    "path": "/tmp/ansible.gdf5vnzz.test/copy/sub/subdir/subdir2/subdir3/subdir4"
                }
            },
            {
                "after": {
                    "path": "/tmp/ansible.gdf5vnzz.test/copy/sub/subdir/subdir1/ansible-test-abs-link",
                    "state": "link"
                },
                "before": {
                    "path": "/tmp/ansible.gdf5vnzz.test/copy/sub/subdir/subdir1/ansible-test-abs-link",
                    "state": "absent"
                }
            },
            {
                "after": {
                    "path": "/tmp/ansible.gdf5vnzz.test/copy/sub/subdir/subdir1/bar.txt",
                    "state": "link"
                },
                "before": {
                    "path": "/tmp/ansible.gdf5vnzz.test/copy/sub/subdir/subdir1/bar.txt",
                    "state": "absent"
                }
            },
            {
                "after": {
                    "path": "/tmp/ansible.gdf5vnzz.test/copy/sub/subdir/subdir1/invalid",
                    "state": "link"
                },
                "before": {
                    "path": "/tmp/ansible.gdf5vnzz.test/copy/sub/subdir/subdir1/invalid",
                    "state": "absent"
                }
            },
            {
                "after": {
                    "path": "/tmp/ansible.gdf5vnzz.test/copy/sub/subdir/subdir1/invalid2",
                    "state": "link"
                },
                "before": {
                    "path": "/tmp/ansible.gdf5vnzz.test/copy/sub/subdir/subdir1/invalid2",
                    "state": "absent"
                }
            },
            {
                "after": {
                    "path": "/tmp/ansible.gdf5vnzz.test/copy/sub/subdir/subdir1/out_of_tree_circle",
                    "state": "link"
                },
                "before": {
                    "path": "/tmp/ansible.gdf5vnzz.test/copy/sub/subdir/subdir1/out_of_tree_circle",
                    "state": "absent"
                }
            },
            {
                "after": {
                    "path": "/tmp/ansible.gdf5vnzz.test/copy/sub/subdir/subdir1/ansible-test-abs-link-dir",
                    "state": "link"
                },
                "before": {
                    "path": "/tmp/ansible.gdf5vnzz.test/copy/sub/subdir/subdir1/ansible-test-abs-link-dir",
                    "state": "absent"
                }
            },
            {
                "after": {
                    "path": "/tmp/ansible.gdf5vnzz.test/copy/sub/subdir/subdir1/circles",
                    "state": "link"
                },
                "before": {
                    "path": "/tmp/ansible.gdf5vnzz.test/copy/sub/subdir/subdir1/circles",
                    "state": "absent"
                }
            },
            {
                "after": {
                    "path": "/tmp/ansible.gdf5vnzz.test/copy/sub/subdir/subdir1/subdir3",
                    "state": "link"
                },
                "before": {
                    "path": "/tmp/ansible.gdf5vnzz.test/copy/sub/subdir/subdir1/subdir3",
                    "state": "absent"
                }
            }
        ],
        "failed": false,
        "src": "/Users/simon/ansible/test/results/.tmp/integration/copy-14vikdc8-ÅÑŚÌβŁÈ/test/integration/targets/copy/files/subdir"
    }
```

</details>

it's clogged up with `before==after` diffs which aren't printed by the default stdout callback plugin. Is this a problem? I could make assertions on the diff elements that aren't printed, but those don't seem very reliable. I wrote some logic to filter these out, but then I ran into a problem where `result['diff']` is not a list after `result.update(execute_module(...))` is called. So I wrote another function `add_module_result_to_my_result`, and used that in place of `result.update`.

But if these `before==after` diffs aren't considered a problem, and asserting that `result.diff | length > 1` is sufficient, then none of that is necessary.

Reply:
I think it would be good to actually check the diff contains the expected contents rather than just the length. It may be easier/clearer to copy a new directory containing a dir/file/symlink (rather than reusing subdir/ and everything it contains).

I looked to see if there are any existing diff mode tests for copy, and there's a single one here https://github.com/ansible/ansible/blob/devel/test/integration/targets/copy/tasks/main.yml#L94-L107. The new test should be added before/after that one so the diff tests are grouped together. (Or else those could go in a new tasks/diff.yml file, since more diff tests will be needed for #84309.)

Reply:
What about copying `content` to a previously-nonexistent `dest`? This should make two diff elements: `state=absent -> state=present` and content.

Reply:
Works for me.

Reply:
ok, that plan didn't work because it's the copy module which should make multiple diffs in that case. Same for changing file permissions. A test that actually applies to my patch is copying *multiple* files, so I went with that.